src/coreclr/nativeaot/Runtime/riscv64/UniversalTransition.S

```diff
@@ -84,88 +86,93 @@
-        fld  fa6, 0x30(a1)         // Load fa6 from a1
-        fld  fa7, 0x38(a1)         // Load fa7 from a1
+        fld fa0, 0x00(a1)
+        fld fa1, 0x08(a1)
+        fld fa2, 0x10(a1)
+        fld fa3, 0x18(a1)
+        fld fa4, 0x20(a1)
+        fld fa5, 0x28(a1)
+        fld fa6, 0x30(a1)
+        fld fa7, 0x38(a1)
 
         PREPARE_EXTERNAL_VAR RhpIntegerTrashValues, a1
 
-        ld  a2, 0x10(a1)           // Load a2 from a1
-        ld  a3, 0x18(a1)           // Load a3 from a1
-        ld  a4, 0x20(a1)           // Load a4 from a1
-        ld  a5, 0x28(a1)           // Load a5 from a1
-        ld  a6, 0x30(a1)           // Load a6 from a1
-        ld  a7, 0x38(a1)           // Load a7 from a1
-    #endif // TRASH_SAVED_ARGUMENT_REGISTERS
-
-        addi a0, sp, DISTANCE_FROM_CHILDSP_TO_RETURN_BLOCK  // First parameter to target function
-        mv a1, t1                                         // Second parameter to target function
-        jalr t0, t1, 0                                    // Jump to the function in t1
-
-    ALTERNATE_ENTRY ReturnFrom\FunctionName
-
-        // Restore the result address from t2
-        mv t2, a0                                       // Move result to t2
-
-        // Restore Floating Point registers
-        fld f0, 0x100(sp)         // Restore f0
-        fld f1, 0x108(sp)         // Restore f1
-        fld f2, 0x110(sp)         // Restore f2
-        fld f3, 0x118(sp)         // Restore f3
-        fld f4, 0x120(sp)         // Restore f4
-        fld f5, 0x128(sp)         // Restore f5
-        fld f6, 0x130(sp)         // Restore f6
-        fld f7, 0x138(sp)         // Restore f7
-
-        // Restore argument registers
-        ld a0, 0x140(sp)         // Restore a0
-        ld a1, 0x148(sp)         // Restore a1
-        ld a2, 0x150(sp)         // Restore a2
-        ld a3, 0x158(sp)         // Restore a3
-        ld a4, 0x160(sp)         // Restore a4
-        ld a5, 0x168(sp)         // Restore a5
-        ld a6, 0x170(sp)         // Restore a6
-        ld a7, 0x178(sp)         // Restore a7
-
-        // Restore FP and RA registers, and free the allocated stack block
-        EPILOG_RESTORE_REG_PAIR_INDEXED  fp, ra, STACK_SIZE
-
-        // Tailcall to the target address in t2
-        jalr t2, 0
+        ld a2, 0x10(a1)
+        ld a3, 0x18(a1)
+        ld a4, 0x20(a1)
+        ld a5, 0x28(a1)
+        ld a6, 0x30(a1)
+        ld a7, 0x38(a1)
+#endif // TRASH_SAVED_ARGUMENT_REGISTERS
+
+        add         a0, sp, DISTANCE_FROM_CHILDSP_TO_RETURN_BLOCK  # First parameter to target function is a pointer to the return block
+        mv          a1, t1                                         # Second parameter to target function
+        jalr        ra, t0, 0
+
+ALTERNATE_ENTRY ReturnFrom\FunctionName
+
+        # Move the result (the target address) to a2 so it doesn't get overridden when we restore the
+        # argument registers.
+        mv          a2, a0
```

Comment:
Should this stay as `t2`? `a2` is overwritten with the `a2` argument below, so the value in a2 won't survive till the end of the method.